r/tools/nixlibs.R

```diff
@@ -96,53 +96,76 @@ try_download <- function(from_url, to_file, hush = quietly) {
-        "Downloading libarrow failed for version %s (%s)\n    at %s",
-        VERSION, lib, binary_url
-      )
-    libfile <- NULL
-  }
+try_checksum <- function(binary_url, libfile, hush = quietly) {
   # Explicitly setting the env var to "false" will skip checksum validation
   # e.g. in case the included checksums are stale.
   skip_checksum <- env_is("ARROW_R_ENFORCE_CHECKSUM", "false")
   enforce_checksum <- env_is("ARROW_R_ENFORCE_CHECKSUM", "true")
   checksum_path <- Sys.getenv("ARROW_R_CHECKSUM_PATH", "tools/checksums")
   # validate binary checksum for CRAN release only
-  if (!skip_checksum && dir.exists(checksum_path) && is_release ||
-    enforce_checksum) {
-    checksum_file <- sub(".+/bin/(.+\\.zip)", "\\1\\.sha512", binary_url)
-    checksum_file <- file.path(checksum_path, checksum_file)
-    checksum_cmd <- "shasum"
-    checksum_args <- c("--status", "-a", "512", "-c", checksum_file)
-
-    # shasum is not available on all linux versions
-    status_shasum <- try(
-      suppressWarnings(
-        system2("shasum", args = c("--help"), stdout = FALSE, stderr = FALSE)
-      ),
-      silent = TRUE
-    )
+  # we do this in a try so that if it fails for any reason, we don't pollute the log
+  # but then we will return that the checksum failed
+  status <- try(
+    {
+      if (!skip_checksum && dir.exists(checksum_path) && is_release ||
+        enforce_checksum) {
+        checksum_file <- sub(".+/bin/(.+\\.zip)", "\\1\\.sha512", binary_url)
+        checksum_file <- file.path(checksum_path, checksum_file)
+        checksum_cmd <- "shasum"
+        checksum_args <- c("--status", "-a", "512", "-c", checksum_file)
+
+        # shasum is not available on all linux versions, so check help if it exists
+        status_shasum <- try(
+          suppressWarnings(
+            system2("shasum", args = c("--help"), stdout = FALSE, stderr = FALSE)
+          ),
+          silent = TRUE
+        )
 
-    if (inherits(status_shasum, "try-error") || is.integer(status_shasum) && status_shasum != 0) {
-      checksum_cmd <- "sha512sum"
-      checksum_args <- c("--status", "-c", checksum_file)
-    }
+        # if shasum doens't exist, then change the command to sha512sum
+        if (inherits(status_shasum, "try-error") || is.integer(status_shasum) && status_shasum != 0) {
+          checksum_cmd <- "sha512sum"
+          checksum_args <- c("--status", "-c", checksum_file)
+        }
 
-    checksum_ok <- system2(checksum_cmd, args = checksum_args)
+        checksum_ok <- system2(checksum_cmd, args = checksum_args)
 
-    if (checksum_ok != 0) {
-      lg("Checksum validation failed for libarrow: %s/%s", lib, libfile)
-      unlink(libfile)
-      libfile <- NULL
-    } else {
-      lg("Checksum validated successfully for libarrow: %s/%s", lib, libfile)
+        if (checksum_ok != 0) {
+          lg("Checksum validation failed for libarrow: %s/%s", lib, libfile)
+          unlink(libfile)
+          stop("The checksum failed")
```

Comment:
I don't think you need these, IIUC `download_binary()` will do the unlinking

```suggestion
```

Reply:
Definitely don't need the stop, but I'm not seeing a place this would be unlinked before we start building (it might get cleaned up in the teardown generally.

I'm going to leave it be for now. `unlink` also _won't error_ if you try and unlink something that exists, which is a bit surprising frankly, but good for us in this case.